gnat2goto/driver/range_check.adb

```diff
@@ -267,12 +264,57 @@ package body Range_Check is
       Value_Expr_Type : constant Irep :=
         Follow_Symbol_Type (Get_Type (Value_Expr), Global_Symbol_Table);
 
-      Op_Geq : constant Irep := New_Irep (I_Op_Geq);
-      Op_Leq : constant Irep := New_Irep (I_Op_Leq);
+      Adjusted_Value_Expr : Irep := Value_Expr;
+      Adjusted_Lower_Bound : Irep := Lower_Bound;
+      Adjusted_Upper_Bound : Irep := Upper_Bound;
+
+      function Set_Value_And_Bounds (Value_Expr : Irep;
+                                     Lower_Bound : Irep;
+                                     Upper_Bound : Irep)
+                                     return Boolean;
+      function Set_Value_And_Bounds (Value_Expr : Irep;
+                                     Lower_Bound : Irep;
+                                     Upper_Bound : Irep)
+                                     return Boolean
+      is
+      begin
+        --  The compared expressions (value and bound) have to be of the
+        --  same type
+         if Get_Width (Bound_Type) > Get_Width (Value_Expr_Type)
+         then
+            --  If the value checked for being in the range is of smaller
+            --  type then we need to cast it to the type of the bounds
+            Adjusted_Value_Expr :=
+              Typecast_If_Necessary
+                (Value_Expr, Bound_Type, Global_Symbol_Table);
+            return True;
+         else
+            --  If the bounds are of smaller type then we cast the bounds
+            --  to the type of the value being checked
+            Adjusted_Lower_Bound :=
+              Typecast_If_Necessary (Lower_Bound, Value_Expr_Type,
+                                     Global_Symbol_Table);
+            Adjusted_Upper_Bound :=
+              Typecast_If_Necessary (Upper_Bound, Value_Expr_Type,
+                                     Global_Symbol_Table);
+            return True;
+         end if;
+      end Set_Value_And_Bounds;
+
+      Values_Adjusted : constant Boolean := Set_Value_And_Bounds
+         (Adjusted_Value_Expr, Adjusted_Lower_Bound, Adjusted_Upper_Bound);
+
+      Op_Geq : constant Irep := Make_Op_Geq (
+        Lhs => (if Values_Adjusted then Adjusted_Value_Expr else Value_Expr),
```

Comment:
But then again why not set lhs to `Adjusted_Value_Expr` (it's initialised to `value_expr`). Same for `rhs`. Then the function above could be a procedure.